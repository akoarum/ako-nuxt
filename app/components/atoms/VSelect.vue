<template>
  <div :class="appearanceClasses" class="select">
    <input
      :id="id ? id : name"
      :name="`${name}-display`"
      :value="displayValue"
      type="text"
      class="select__input"
      readonly
      @click="openOption"
    />
    <transition name="option">
      <div v-show="isOpened" class="select__optionContainer">
        <ul class="select__options">
          <li v-for="option in options" :key="option.id">
            <label>
              <input
                type="radio"
                :name="name"
                :value="option.value"
                :checked="value === option.value"
                :required="required"
                @change="selectOption"
              />
              <span>{{ option.display || option.value }}</span>
            </label>
          </li>
        </ul>
      </div>
    </transition>
    <div v-if="isOpened" class="select__mask" @click="closeOption" />
  </div>
</template>

<script>
export default {
  props: {
    name: { type: String, required: true },
    value: { type: String, required: true },
    options: { type: Array, required: true },
    placeholder: { type: String, default: '選択してください' },
    id: { type: String },
    error: { type: Boolean },
    required: { type: Boolean }
  },
  data() {
    return {
      isOpened: false
    }
  },
  computed: {
    displayValue() {
      if (!this.value) return this.placeholder

      for (const option of this.options) {
        if (this.value !== String(option.value)) continue
        return option.display || option.value
      }

      return this.placeholder
    },
    appearanceClasses() {
      const classes = []

      if (this.error) {
        classes.push('-error')
      }

      return classes
    }
  },
  methods: {
    openOption() {
      this.isOpened = true
      window.addEventListener('keyup', this.escKeyHandler)
    },
    closeOption() {
      this.isOpened = false
      window.removeEventListener('keyup', this.escKeyHandler)
    },
    selectOption(event) {
      this.$emit('change', event.target.value)
      this.closeOption()
    },
    escKeyHandler(event) {
      if (event.keyCode !== 27) return
      this.closeOption()
    }
  }
}
</script>

<style lang="scss" scoped>
.select {
  position: relative;
  display: inline-block;

  &__input {
    box-sizing: border-box;
    border-radius: 4px;
    background: $color-inputBack;
    font-size: $fontSize-base;
    outline: none;
    transition: box-shadow $transition-duration $transition-easing,
      background $transition-duration $transition-easing;
    will-change: box-shadow, background;
    cursor: pointer;
  }

  &__optionContainer {
    position: absolute;
    left: 0;
    top: 100%;
    border-radius: 4px;
    width: 100%;
    background: #fff;
    z-index: 20;
  }
  &__options {
    border: 1px solid $color-border;

    > li {
      position: relative;

      + li {
        border-top: 1px solid $color-border;
      }
    }
    input {
      position: absolute;
      left: 0;
      top: 0;
      opacity: 0;
      width: 1px;
      height: 1px;

      &:checked + span {
        background: $color-inputBack;
      }
    }
    label {
      display: block;
      cursor: pointer;

      &:hover,
      &:focus {
        span {
          background: $color-inputBack;
        }
      }
    }
    span {
      display: block;
      transition: background $transition-duration $transition-easing;
      will-change: background;
    }
  }
  &__mask {
    position: fixed;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    z-index: 19;
  }

  &.-error {
    .select__input {
      background: #ffe8e8;
    }
  }

  @media screen and (min-width: 768px) {
    &__input {
      padding: 4px 12px;
    }
    &__options {
      label {
        font-size: $fontSize-base;
      }
      span {
        padding: 7px 12px;
      }
    }
  }

  @media screen and (max-width: 767px) {
    display: block;

    &__input {
      padding: 7px 12px;
      width: 100%;
    }
    &__options {
      label {
        font-size: $fontSize-sp;
      }
      span {
        padding: 9px 14px;
      }
    }
  }
}

.option {
  &-enter-active,
  &-leave-active {
    transition: opacity $transition-duration $transition-easing;
    will-change: opacity;
  }
  &-enter,
  &-leave-to {
    opacity: 0;
  }
  &-enter-to,
  &-leave {
    opacity: 1;
  }
}
</style>
