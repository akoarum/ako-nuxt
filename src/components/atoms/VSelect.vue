<template>
  <div class="vSelect">
    <label class="vSelect__label" @click="updateOptionView(!optionVisible)">
      {{ selectedView }}
      <VCaret :width="10" :height="6" class="vSelect__icon" />
    </label>
    <transition name="option">
      <ul v-show="optionVisible" class="vSelect__options">
        <li v-for="option in options" :key="option.id" class="option">
          <label class="option__label">
            <input
              v-bind="$attrs"
              :value="option.value"
              :checked="isSelected(option.value)"
              type="radio"
              class="option__input"
              @change="$emit('change', option.value)"
              @click="updateOptionView(false)"
            />
            <span class="option__txt">{{ option.label }}</span>
          </label>
        </li>
      </ul>
    </transition>
  </div>
</template>

<script>
import { VCaret } from '~/components/icons'

export default {
  components: { VCaret },
  inheritAttrs: false,
  model: {
    event: 'change',
    prop: 'checkedValue'
  },
  props: {
    options: { type: Array, required: true },
    checkedValue: { type: String }
  },
  data() {
    return {
      optionVisible: false
    }
  },
  computed: {
    selectedView() {
      if (!this.checkedValue) return '選択してください'
      return this.options.find(option => option.value === this.checkedValue)
        .label
    }
  },
  methods: {
    isSelected(value) {
      return this.checkedValue === value
    },
    updateOptionView(status) {
      this.optionVisible = status
    }
  }
}
</script>

<style lang="scss" scoped>
.vSelect {
  position: relative;
  display: block;
}

.vSelect__label {
  display: block;
  box-sizing: border-box;
  border: 1px solid $color-border;
  padding: 8px 12px;
  font-size: $fontSize-pc;
  outline: none;
  transition: border 0.3s $easing;
}

.vSelect__icon {
  position: absolute;
  right: 12px;
  top: 50%;
  transform: rotate(90deg);
}

.vSelect__options {
  position: absolute;
  left: 13px;
  top: 90%;
  box-shadow: 0 3px 8px 0 rgba(100, 100, 100, 0.4);
  min-width: 120px;
  background: #fff;
}

.option {
  position: relative;
}

.option__input {
  position: absolute;
  left: 0;
  top: 0;
  width: 5px;
  height: 5px;
  opacity: 0;

  &:checked + .option__txt {
    background: #efefef;
  }
}

.option__label {
  display: block;
  transition: background 0.3s $easing;

  &:hover,
  &:focus {
    background: #efefef;
  }
}

.option__txt {
  display: block;
  padding: 12px;
  font-size: $fontSize-pc;
  line-height: 1;
}

.option {
  &-enter-active,
  &-leave-active {
    transition: opacity 0.3s $easing;
  }
  &-enter,
  &-leave-to {
    opacity: 0;
  }
}
</style>
