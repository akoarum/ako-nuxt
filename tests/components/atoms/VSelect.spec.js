import { shallowMount } from '@vue/test-utils'
import VSelect from '~/components/atoms/VSelect'

describe('VSelect', () => {
  let wrapper

  const options = [
    { id: 1, label: '項目1', value: '1' },
    { id: 2, label: '項目2', value: '2' },
    { id: 3, label: '項目3', value: '3' }
  ]

  beforeEach(() => {
    wrapper = shallowMount(VSelect, {
      propsData: {
        options,
        checkedValue: ''
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })
    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('computed', () => {
    describe('selectedView', () => {
      it('選択されているoptionのlabelを表示する', () => {
        expect(VSelect.computed.selectedView.call({ options, checkedValue: '2' })).toBe('項目2')
      })

      it('選択されていない場合は「選択してください」を表示する', () => {
        expect(VSelect.computed.selectedView.call({ options, checkedValue: '' })).toBe('選択してください')
      })
    })
  })

  describe('methods', () => {
    it('isSelected: 引数valueとcheckedValueを比較してcheckedかどうかを返す', () => {
      wrapper.setProps({ checkedValue: '1' })
      expect(wrapper.vm.isSelected('1')).toBe(true)
      expect(wrapper.vm.isSelected('2')).toBe(false)
    })

    it('updateOptionView: 引数statusをoptionVisibleに入れる', () => {
      wrapper.vm.updateOptionView(true)
      expect(wrapper.vm.optionVisible).toBe(true)

      wrapper.vm.updateOptionView(false)
      expect(wrapper.vm.optionVisible).toBe(false)
    })
  })
})
