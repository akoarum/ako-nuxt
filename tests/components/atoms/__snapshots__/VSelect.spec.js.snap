// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VSelect initialized snapshot 1`] = `
<div
  class="vSelect"
>
  <label
    class="vSelect__label"
  >
    
    選択してください
    
    <vcaret-stub
      class="vSelect__icon"
      height="6"
      width="10"
    />
  </label>
   
  <ul
    class="vSelect__options"
    name="option"
    style="display: none;"
  >
    <li
      class="option"
    >
      <label
        class="option__label"
      >
        <input
          class="option__input"
          type="radio"
          value="1"
        />
         
        <span
          class="option__txt"
        >
          項目1
        </span>
      </label>
    </li>
    <li
      class="option"
    >
      <label
        class="option__label"
      >
        <input
          class="option__input"
          type="radio"
          value="2"
        />
         
        <span
          class="option__txt"
        >
          項目2
        </span>
      </label>
    </li>
    <li
      class="option"
    >
      <label
        class="option__label"
      >
        <input
          class="option__input"
          type="radio"
          value="3"
        />
         
        <span
          class="option__txt"
        >
          項目3
        </span>
      </label>
    </li>
  </ul>
</div>
`;
